File: r_sql/construct_database.R
library(DBI)
library(RSQLite)
library(readr)

# === SET YOUR DIRECTORY CONTAINING THE .bz2 AND .csv FILES ===
data_dir <- "~/Desktop/st_2195/st2195_assignment_3_database/dataverse_files"

# === CREATE DATABASE in same folder ===
db_path <- file.path(data_dir, "airline2.db")
con <- dbConnect(SQLite(), db_path)

# === Get .csv.bz2 files for 2000–2005 ===
files <- list.files(path = data_dir, pattern = "^200[6-8]\\.csv\\.bz2$", full.names = TRUE)

# === Read and append into "ontime" table ===
for (file in files) {
  message("Processing: ", file)
  df <- read_csv(file)
  dbWriteTable(con, "ontime", df, append = TRUE)
}

# === Read and write lookup tables ===
airports <- read_csv(file.path(data_dir, "airports.csv"))
dbWriteTable(con, "airports", airports, overwrite = TRUE)

carriers <- read_csv(file.path(data_dir, "carriers.csv"))
dbWriteTable(con, "carriers", carriers, overwrite = TRUE)

planes <- read_csv(file.path(data_dir, "plane-data.csv"))
dbWriteTable(con, "planes", planes, overwrite = TRUE)

dbDisconnect(con)
cat("Database successfully created at:", db_path, "\n")

File: python_sql/construct_database.py
import os
import pandas as pd
import sqlite3

# === Set your directory paths ===
data_dir = "/Users/joanneodannell/Desktop/st_2195/st2195_assignment_3_database/dataverse_files"  # change this
db_path = os.path.join(data_dir, "airline.db")

# === Connect to SQLite database ===
conn = sqlite3.connect(db_path)
cursor = conn.cursor()

# === Process each .bz2 file in chunks ===
years = range(2006,2009)
for year in years:
    file_path = os.path.join(data_dir, f"{year}.csv.bz2")
    print(f"Processing {file_path}")

    chunk_iter = pd.read_csv(file_path, chunksize=100000)  # memory safe

    for i, chunk in enumerate(chunk_iter):
        chunk.to_sql("ontime", conn, if_exists="append", index=False)
        print(f"  → Chunk {i+1} written")

# === Load reference tables ===
for name, filename in [
    ("airports", "airports.csv"),
    ("carriers", "carriers.csv"),
    ("planes", "plane-data.csv")
]:
    print(f"Loading {name}")
    df = pd.read_csv(os.path.join(data_dir, filename))
    df.to_sql(name, conn, if_exists="replace", index=False)

# === Done ===
conn.close()
print(f"Database created at {db_path}")

